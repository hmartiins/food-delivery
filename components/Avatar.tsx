import { Image, TouchableOpacity, View } from 'react-native';

import { images } from '@/constants';

type AvatarProps = {
  onPress?: () => void;
  hasEdit?: boolean;
};

export const Avatar = ({
  hasEdit = false,
  onPress = () => {},
}: AvatarProps) => {
  return (
    <TouchableOpacity
      testID="avatar-container"
      className="relative"
      onPress={hasEdit ? onPress : undefined}
    >
      <Image
        testID="avatar-image"
        source={{
          uri: 'https://github.com/hmartiins.png',
        }}
        className="size-24 rounded-full"
        resizeMode="contain"
      />

      {hasEdit && (
        <View
          testID="avatar-edit-badge"
          className="absolute bottom-0 right-0 size-7 rounded-full border border-white-100 bg-primary"
        >
          <Image
            testID="avatar-edit-icon"
            source={images.pencil}
            className="mx-auto my-auto mt-auto size-3"
            resizeMode="contain"
          />
        </View>
      )}
    </TouchableOpacity>
  );
};
