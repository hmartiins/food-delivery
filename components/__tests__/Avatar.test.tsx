import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react-native';

import { Avatar } from '../Avatar';

describe('<Avatar />', () => {
  describe('Rendering', () => {
    it('should render without error', () => {
      render(<Avatar />);
      expect(screen.getByTestId('avatar-container')).toBeTruthy();
      expect(screen.getByTestId('avatar-image')).toBeTruthy();
    });

    it('should display avatar image correctly', () => {
      render(<Avatar />);
      const avatarImage = screen.getByTestId('avatar-image');

      expect(avatarImage).toBeTruthy();
      expect(avatarImage.props.source.uri).toBe(
        'https://github.com/hmartiins.png'
      );
    });
  });

  describe('Edit badge behavior', () => {
    it('should not display edit badge when hasEdit is false', () => {
      render(<Avatar hasEdit={false} />);

      expect(screen.getByTestId('avatar-container')).toBeTruthy();
      expect(screen.queryByTestId('avatar-edit-badge')).toBeNull();
      expect(screen.queryByTestId('avatar-edit-icon')).toBeNull();
    });

    it('should display edit badge when hasEdit is true', () => {
      render(<Avatar hasEdit={true} />);

      expect(screen.getByTestId('avatar-container')).toBeTruthy();
      expect(screen.getByTestId('avatar-edit-badge')).toBeTruthy();
      expect(screen.getByTestId('avatar-edit-icon')).toBeTruthy();
    });

    it('should not display edit badge by default', () => {
      render(<Avatar />);

      expect(screen.queryByTestId('avatar-edit-badge')).toBeNull();
      expect(screen.queryByTestId('avatar-edit-icon')).toBeNull();
    });
  });

  describe('Interactions', () => {
    it('should call onPress when hasEdit is true and avatar is pressed', () => {
      const mockOnPress = jest.fn();
      render(<Avatar hasEdit={true} onPress={mockOnPress} />);

      const avatarContainer = screen.getByTestId('avatar-container');
      fireEvent.press(avatarContainer);

      expect(mockOnPress).toHaveBeenCalledTimes(1);
    });

    it('should allow multiple presses when hasEdit is true', () => {
      const mockOnPress = jest.fn();
      render(<Avatar hasEdit={true} onPress={mockOnPress} />);

      const avatarContainer = screen.getByTestId('avatar-container');
      fireEvent.press(avatarContainer);
      fireEvent.press(avatarContainer);
      fireEvent.press(avatarContainer);

      expect(mockOnPress).toHaveBeenCalledTimes(3);
    });
  });
});

describe('<Avatar /> Snapshots', () => {
  it('should match snapshot with default props', () => {
    const component = render(<Avatar />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match snapshot with hasEdit false', () => {
    const component = render(<Avatar hasEdit={false} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match snapshot with hasEdit true', () => {
    const component = render(<Avatar hasEdit={true} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
