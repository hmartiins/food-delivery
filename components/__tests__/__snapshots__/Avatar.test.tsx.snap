// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> Snapshots should match snapshot with default props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="avatar-container"
>
  <Image
    className="size-24 rounded-full"
    resizeMode="contain"
    source={
      {
        "uri": "https://github.com/hmartiins.png",
      }
    }
    testID="avatar-image"
  />
</View>
`;

exports[`<Avatar /> Snapshots should match snapshot with hasEdit false 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="avatar-container"
>
  <Image
    className="size-24 rounded-full"
    resizeMode="contain"
    source={
      {
        "uri": "https://github.com/hmartiins.png",
      }
    }
    testID="avatar-image"
  />
</View>
`;

exports[`<Avatar /> Snapshots should match snapshot with hasEdit true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="avatar-container"
>
  <Image
    className="size-24 rounded-full"
    resizeMode="contain"
    source={
      {
        "uri": "https://github.com/hmartiins.png",
      }
    }
    testID="avatar-image"
  />
  <View
    className="absolute bottom-0 right-0 size-7 rounded-full border border-white-100 bg-primary"
    testID="avatar-edit-badge"
  >
    <Image
      className="mx-auto my-auto mt-auto size-3"
      resizeMode="contain"
      source={
        {
          "testUri": "../../../assets/icons/pencil.png",
        }
      }
      testID="avatar-edit-icon"
    />
  </View>
</View>
`;
